ompi/mca/osc/rdma/osc_rdma_passive_target.c

```diff
@@ -263,6 +339,63 @@ int ompi_osc_rdma_lock_atomic (int lock_type, int target, int assert, ompi_win_t
     return ret;
 }
 
+int ompi_osc_rdma_ilock_atomic (int lock_type, int target, int assert, ompi_win_t *win)
+{
+    ompi_osc_rdma_module_t *module = GET_MODULE(win);
+    ompi_osc_rdma_peer_t *peer = ompi_osc_rdma_module_peer (module, target);
+    ompi_osc_rdma_sync_t *lock;
+    int ret = OMPI_SUCCESS;
+
+    OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_TRACE, "lock: %d, %d, %d, %s", lock_type, target, assert, win->w_name);
+
+    if (module->no_locks) {
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "attempted to lock with no_locks set");
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    if (module->all_sync.epoch_active && (OMPI_OSC_RDMA_SYNC_TYPE_LOCK != module->all_sync.type || MPI_LOCK_EXCLUSIVE == lock_type)) {
+        /* impossible to get an exclusive lock while holding a global shared lock or in a active
+         * target access epoch */
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    /* clear the global sync object (in case MPI_Win_fence was called) */
+    module->all_sync.type = OMPI_OSC_RDMA_SYNC_TYPE_NONE;
+
+    /* create lock item */
+    lock = ompi_osc_rdma_sync_allocate (module);
+    if (OPAL_UNLIKELY(NULL == lock)) {
+        return OMPI_ERR_OUT_OF_RESOURCE;
+    }
+
+    lock->type = OMPI_OSC_RDMA_SYNC_TYPE_LOCK;
+    lock->sync.lock.target = target;
+    lock->sync.lock.type = lock_type;
+    lock->sync.lock.assert = assert;
+
+    lock->peer_list.peer = peer;
+    lock->num_peers = 1;
+    OBJ_RETAIN(peer);
+
+    if (0 == (assert & MPI_MODE_NOCHECK)) {
+        ret = ompi_osc_rdma_ilock_atomic_internal (module, peer, lock);
+    }
+
+    if (OPAL_LIKELY(OMPI_SUCCESS == ret)) {
+        ++module->passive_target_access_epoch;
+
+        opal_atomic_wmb ();
+
+        OPAL_THREAD_SCOPED_LOCK(&module->lock, ompi_osc_rdma_module_lock_insert (module, lock));
+    } else {
+        OBJ_RELEASE(lock);
```

Comment:
I'm worried this won't release the reference to `peer` taken above in `OBJ_RETAIN(peer)`.

The `iunlock` code does:
```
OBJ_RELEASE(peer);
ompi_osc_rdma_sync_return (lock);
```
when it wants to destroy the `lock` object.

Reply:
Added OBJ_RELEASE(peer) to release reference to peer